File: src/lib/backoff.js
"use strict";

const MIN_MS = 15_000;
const MAX_MS = 15 * 60 * 1000;

function computeBackoffMs(consecutiveFailures) {
  const n = Number(consecutiveFailures) || 0;
  if (n <= 0) return 0;
  return Math.min(MAX_MS, MIN_MS * 2 ** (n - 1));
}

function createBackoffTracker() {
  const failures = new Map();
  const nextAllowedAt = new Map();

  return {
    recordSuccess(id) {
      failures.delete(id);
      nextAllowedAt.delete(id);
    },
    recordFailure(id, now = Date.now()) {
      const count = (failures.get(id) || 0) + 1;
      failures.set(id, count);
      nextAllowedAt.set(id, now + computeBackoffMs(count));
      return computeBackoffMs(count);
    },
    shouldSkip(id, now = Date.now()) {
      const until = nextAllowedAt.get(id);
      return typeof until === "number" && now < until;
    },
    consecutiveFailures(id) {
      return failures.get(id) || 0;
    },
  };
}

module.exports = { computeBackoffMs, createBackoffTracker, MIN_MS, MAX_MS };

File: QuotaHud/Domain/Backoff.cs
namespace QuotaHud;

public sealed class BackoffTracker
{
    public static readonly TimeSpan Min = TimeSpan.FromSeconds(15);
    public static readonly TimeSpan Max = TimeSpan.FromMinutes(15);
    readonly Dictionary<string, int> _failures = new();
    readonly Dictionary<string, DateTime> _nextAllowedAt = new();

    public static TimeSpan Compute(int consecutiveFailures)
    {
        if (consecutiveFailures <= 0) return TimeSpan.Zero;
        var ms = Min.TotalMilliseconds * Math.Pow(2, consecutiveFailures - 1);
        return TimeSpan.FromMilliseconds(Math.Min(Max.TotalMilliseconds, ms));
    }

    public void RecordSuccess(string id)
    {
        _failures.Remove(id);
        _nextAllowedAt.Remove(id);
    }

    public TimeSpan RecordFailure(string id, DateTime? now = null)
    {
        var count = (_failures.TryGetValue(id, out var n) ? n : 0) + 1;
        _failures[id] = count;
        var delay = Compute(count);
        _nextAllowedAt[id] = (now ?? DateTime.UtcNow) + delay;
        return delay;
    }

    public bool ShouldSkip(string id, DateTime? now = null)
    {
        if (!_nextAllowedAt.TryGetValue(id, out var until)) return false;
        return (now ?? DateTime.UtcNow) < until;
    }
}
